std-bits/table/src/main/java/org/enso/table/parsing/NumberParser.java

```diff
@@ -196,259 +26,132 @@ private static Pattern buildPattern(
-
-    if (separatorsIndex >= SEPARATORS.length) {
-      return null;
-    }
-
-    return buildPattern(
-        allowDecimal,
-        allowCurrency,
-        allowScientific,
-        trimValues,
-        patternIndex,
-        SEPARATORS[separatorsIndex]);
+    var numberWithSeparators = NumberWithSeparators.fromSeparators(thousandSeparator, decimalPoint);
+    this.parser =
+        new FormatDetectingNumberParser(
+            allowSymbol,
+            allowLeadingZeroes,
+            allowLeadingTrailingWhitespace,
+            NegativeSign.UNKNOWN,
+            numberWithSeparators);
   }
 
   @Override
-  public Object parseSingleValue(String text, ParseProblemAggregator problemAggregator) {
-    int index = 0;
-    var pattern = patternForIndex(index);
-    while (pattern != null) {
-      var value = innerParseSingleValue(text, pattern);
-      if (value != null) {
-        return value;
-      }
-
-      index++;
-      pattern = patternForIndex(index);
-    }
-
-    problemAggregator.reportInvalidFormat(text);
-    return null;
+  protected Builder makeBuilderWithCapacity(int capacity, ProblemAggregator problemAggregator) {
+    return isInteger
+        ? NumericBuilder.createLongBuilder(capacity, integerTargetType, problemAggregator)
+        : NumericBuilder.createDoubleBuilder(capacity, problemAggregator);
   }
 
   @Override
   public Storage<?> parseColumn(
       Storage<String> sourceStorage, CommonParseProblemAggregator problemAggregator) {
-    int index = 0;
-    var pattern = patternForIndex(index);
-
-    int bestIndex = 0;
-    int bestCount = -1;
-    while (pattern != null) {
-      ProblemAggregator inner = problemAggregator.createSimpleChild();
-      Builder builder = makeBuilderWithCapacity(sourceStorage.size(), inner);
-      int failedAt = parseColumnWithPattern(pattern, sourceStorage, builder, null);
-      if (failedAt == -1) {
-        return builder.seal();
-      }
+    Builder builder =
+        makeBuilderWithCapacity(sourceStorage.size(), problemAggregator.createSimpleChild());
 
-      // If there was a failure, we abandon this branch - thus we discard any problems that might
-      // have been reported by the inner aggregator.
-      inner.detachFromParent();
+    var context = Context.getCurrent();
+    for (int i = 0; i < sourceStorage.size(); i++) {
+      var text = sourceStorage.getItemBoxed(i);
 
-      if (failedAt > bestCount) {
-        bestCount = failedAt;
-        bestIndex = index;
+      // Check if in unknown state
+      var isInMixedState =
+          !isInteger
+              && (parser.numberWithSeparators() == NumberWithSeparators.DOT_UNKNOWN
+                  || parser.numberWithSeparators() == NumberWithSeparators.COMMA_UNKNOWN);
+
+      // Try and parse the value
+      var result = text == null ? null : parseSingleValue(text, problemAggregator);
+
+      // Do we need to rescan?
+      if (!isInMixedState && parser.numberWithSeparators() != NumberWithSeparators.DOT_COMMA) {
```

Comment:
Why `parser.numberWithSeparators() != NumberWithSeparators.DOT_COMMA` is a condition for rescan? It's not immediately clear to me, can we get a comment explaining the reasoning here?

Reply:
Then, I'd probably also replace this with a higher-level getter on the parser that will make this condition easier to understand by giving it a meaningful name.

```suggestion
      if (!isInMixedState && parser.someConditionReached()) {
```

Not sure what `someConditionReached` should be called yet because the intent is unclear (for me) from current code.